VideoApi/VideoApi/Services/VirtualRoomService.cs

```diff
@@ -0,0 +1,103 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using VideoApi.DAL.Commands;
+using VideoApi.DAL.Commands.Core;
+using VideoApi.DAL.Queries;
+using VideoApi.DAL.Queries.Core;
+using VideoApi.Domain;
+using VideoApi.Domain.Enums;
+using VideoApi.Services.Contracts;
+using VideoApi.Services.Kinly;
+
+namespace VideoApi.Services
+{
+    public class VirtualRoomService : IVirtualRoomService
+    {
+        private readonly IKinlyApiClient _kinlyApiClient;
+        private readonly ILogger<VirtualRoomService> _logger;
+        private readonly ICommandHandler _commandHandler;
+        private readonly IQueryHandler _queryHandler;
+
+        public VirtualRoomService(IKinlyApiClient kinlyApiClient, ILogger<VirtualRoomService> logger,
+            ICommandHandler commandHandler, IQueryHandler queryHandler)
+        {
+            _kinlyApiClient = kinlyApiClient;
+            _logger = logger;
+            _commandHandler = commandHandler;
+            _queryHandler = queryHandler;
+        }
+
+        public async Task<Room> GetOrCreateAnInterpreterVirtualRoom(Conference conference, Participant participant)
+        {
+            var conferenceId = conference.Id;
+
+            // get a room
+            var roomPrefix = "Interpreter";
+            var query = new GetAvailableRoomByRoomTypeQuery(VirtualCourtRoomType.Civilian, conferenceId);
+            var listOfRooms = await _queryHandler.Handle<GetAvailableRoomByRoomTypeQuery, List<Room>>(query);
+            var interpreterRooms = listOfRooms.Where(x => x.Label.Contains(roomPrefix)).ToList();
+            var room = GetRoomForParticipant(conferenceId, interpreterRooms, participant) ??
+                       await CreateAnInterpreterRoom(conference, interpreterRooms.Count);
+
+            return room;
+        }
+
+        private async Task<Room> CreateAnInterpreterRoom(Conference conference, int existingRooms)
+        {
+            _logger.LogInformation("Creating a new interpreter room for conference {Conference}", conference.Id);
+            var createRoomCommand = new CreateRoomCommand(conference.Id, null, VirtualCourtRoomType.Civilian, false);
+            await _commandHandler.Handle(createRoomCommand);
+            var roomId = createRoomCommand.NewRoomId;
+            _logger.LogInformation("Creating a new civilian room {Room} for conference {Conference}", roomId,
+                conference.Id);
+            var ingestUrl = $"{conference.IngestUrl}/{roomId}";
+            var roomPrefix = "Interpreter";
+
+            var newRoomParams = new CreateParticipantRoomParams
+            {
+                Audio_recording_url = ingestUrl,
+                Participant_room_id = roomId.ToString(),
+                Participant_type = VirtualCourtRoomType.Civilian.ToString(),
+                Room_label_prefix = roomPrefix,
+                Room_type = roomPrefix,
+                Display_name = $"{roomPrefix}{existingRooms + 1}",
```

Comment:
we're incrementing? not kinly?

Reply:
This one is a little tricky. I've raised this and I believe Kinly are going to make changes so they produce a label like they do with the other rooms